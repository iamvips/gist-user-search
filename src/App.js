import { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import styled from "styled-components";
import ReactPaginate from "react-paginate";
import Header from "./components/Header";
import GlobalStyles from "./GlobalStyle";
import { getPublicGists } from "./services/gistService";
import GistList from "./components/GistList";
import NoResults from "./components/NoResults";
import Footer from "./components/Footer";
import {
  searchUserGists,
  setUserGistResults,
  searchUserGistsFailure,
  setPageNumber,
} from "./redux/actions/searchAction";
import ErrorPage from "./components/ErrorPage";
import ErrorBoundary from "./components/ErrorBoundary";

const App = () => {
  const { gists, pageNumber, isError } = useSelector(
    (state) => state.searchReducer
  );
  const dispatch = useDispatch();

  const gistsPerPage = 8;
  const pagesVisited = pageNumber * gistsPerPage;
  const pageCount = Math.ceil(gists.length / gistsPerPage);
  const changePage = ({ selected }) => {
    dispatch(setPageNumber(selected));
  };

  const paginatedResults = gists.slice(
    pagesVisited,
    pagesVisited + gistsPerPage
  );

  /**
   * Displays all public gists on initial load
   */
  useEffect(() => {
    fetchAllPublicGists();

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const fetchAllPublicGists = () => {
    dispatch(searchUserGists());
    getPublicGists()
      .then((res) => {
        dispatch(setUserGistResults(res.data));
      })
      .catch((err) => {
        console.log(err);
        dispatch(searchUserGistsFailure(err.message));
      });
  };

  return (
    <Wrapper className="App" data-testid="app">
      <Header onClick={fetchAllPublicGists} />
      <GlobalStyles />
      <ErrorBoundary>
        {paginatedResults.length < 1 && !isError && <NoResults />}

        {isError && <ErrorPage />}

        {paginatedResults && <GistList gists={paginatedResults} />}

        {paginatedResults && paginatedResults.length > 0 && (
          <ReactPaginate
            previousLabel={"Previous"}
            nextLabel={"Next"}
            pageCount={pageCount}
            onPageChange={changePage}
            containerClassName={"paginationBttns"}
            previousLinkClassName={"previousBttn"}
            nextLinkClassName={"nextBttn"}
            disabledClassName={"paginationDisabled"}
            activeClassName={"paginationActive"}
            forcePage={pageNumber}
          />
        )}
      </ErrorBoundary>
      <Footer />
    </Wrapper>
  );
};

const Wrapper = styled.div`
  font-size: 14px;
  line-height: 1.5;
`;

export default App;
