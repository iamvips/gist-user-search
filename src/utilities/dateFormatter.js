/**
 * @param stringDate
 * @returns Formated date in mm/dd/yyyy format
 */
export const formatDate = (stringDate) => {
  const formattedDate = new Date(stringDate);
  const date = formattedDate.toLocaleString("en-us").split(",")[0];
  return date;
};
