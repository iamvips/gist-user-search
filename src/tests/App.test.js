import React from "react";
import renderer from "react-test-renderer";
import { render, fireEvent, screen } from "../utilities/test-utils";
import App from "../App";
import * as redux from "react-redux";

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe("App Component", () => {
  let spyOnUseSelector;
  let spyOnUseDispatch;
  let mockDispatch;
  const gistsMock = [
    {
      url: "https://api.github.com/gistsMock/45a4808a6c743ff4ce630276845cee97",
      forks_url:
        "https://api.github.com/gistsMock/45a4808a6c743ff4ce630276845cee97/forks",
      commits_url:
        "https://api.github.com/gistsMock/45a4808a6c743ff4ce630276845cee97/commits",
      id: "45a4808a6c743ff4ce630276845cee97",
      node_id: "MDQ6R2lzdDQ1YTQ4MDhhNmM3NDNmZjRjZTYzMDI3Njg0NWNlZTk3",
      git_pull_url:
        "https://gist.github.com/45a4808a6c743ff4ce630276845cee97.git",
      git_push_url:
        "https://gist.github.com/45a4808a6c743ff4ce630276845cee97.git",
      html_url: "https://gist.github.com/45a4808a6c743ff4ce630276845cee97",
      files: {
        "crash.txt": {
          filename: "crash.txt",
          type: "text/plain",
          language: "Text",
          raw_url:
            "https://gist.githubusercontent.com/natanfudge/45a4808a6c743ff4ce630276845cee97/raw/25baccdb12b105d5889158ddfb0a104d06c7e045/crash.txt",
          size: 42772,
        },
      },
      public: true,
      created_at: "2021-04-17T03:22:06Z",
      updated_at: "2021-04-17T03:22:07Z",
      description: null,
      comments: 0,
      user: null,
      comments_url:
        "https://api.github.com/gistsMock/45a4808a6c743ff4ce630276845cee97/comments",
      owner: {
        login: "natanfudge",
        id: 25990014,
        node_id: "MDQ6VXNlcjI1OTkwMDE0",
        avatar_url: "https://avatars.githubusercontent.com/u/25990014?v=4",
        gravatar_id: "",
        url: "https://api.github.com/users/natanfudge",
        html_url: "https://github.com/natanfudge",
        followers_url: "https://api.github.com/users/natanfudge/followers",
        following_url:
          "https://api.github.com/users/natanfudge/following{/other_user}",
        gistsMock_url:
          "https://api.github.com/users/natanfudge/gistsMock{/gist_id}",
        starred_url:
          "https://api.github.com/users/natanfudge/starred{/owner}{/repo}",
        subscriptions_url:
          "https://api.github.com/users/natanfudge/subscriptions",
        organizations_url: "https://api.github.com/users/natanfudge/orgs",
        repos_url: "https://api.github.com/users/natanfudge/repos",
        events_url: "https://api.github.com/users/natanfudge/events{/privacy}",
        received_events_url:
          "https://api.github.com/users/natanfudge/received_events",
        type: "User",
        site_admin: false,
      },
      truncated: false,
    },
  ];
  const mockState = { gists: gistsMock };

  beforeEach(() => {
    // Mock useSelector hook
    spyOnUseSelector = jest.spyOn(redux, "useSelector");
    spyOnUseSelector.mockReturnValue(mockState);
    // Mock useDispatch hook
    spyOnUseDispatch = jest.spyOn(redux, "useDispatch");
    // Mock dispatch function returned from useDispatch
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);
  });

  afterEach(() => {
    spyOnUseSelector.mockClear();
    spyOnUseDispatch.mockClear();
  });

  it("should render correctly", () => {
    const { queryByTestId } = render(<App />, mockState);
    const appComponent = queryByTestId("app");
    expect(appComponent).toBeTruthy();
  });

  test("renders learn react link", () => {
    render(<App />);
    const linkElement = screen.getByTestId("app");
    expect(linkElement).toBeInTheDocument();
  });

  it("should render and match snapshot", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
