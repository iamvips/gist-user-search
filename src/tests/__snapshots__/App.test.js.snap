// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render and match snapshot 1`] = `
<div
  className="sc-bkkeKt grtncx App"
  data-testid="app"
>
  <div
    className="sc-hKwDye bUKynq"
    data-testid="header"
  >
    <span
      className="mega-octicon octicon-mark-github"
    />
    <div
      className="sc-bdvvtL jcZpGJ"
    >
      <div
        className="sc-gsDKAQ ldqIrX"
      >
        <span
          className="octicon octicon-search"
        />
        <input
          className="sc-dkPtRN lpkptG"
          data-testid="search-input"
          onChange={[Function]}
          placeholder="Search Gists for the username"
        />
        <div
          aria-busy="true"
          className=""
          style={Object {}}
        >
          <svg
            aria-label="audio-loading"
            height={25}
            stroke="#00BFFF"
            viewBox="0 0 44 44"
            width={25}
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fillRule="evenodd"
              strokeWidth="2"
            >
              <circle
                cx="22"
                cy="22"
                r={1}
              >
                <animate
                  attributeName="r"
                  begin="0s"
                  calcMode="spline"
                  dur="1.8s"
                  keySplines="0.165, 0.84, 0.44, 1"
                  keyTimes="0; 1"
                  repeatCount="indefinite"
                  values="1; 20"
                />
                <animate
                  attributeName="strokeOpacity"
                  begin="0s"
                  calcMode="spline"
                  dur="1.8s"
                  keySplines="0.3, 0.61, 0.355, 1"
                  keyTimes="0; 1"
                  repeatCount="indefinite"
                  values="1; 0"
                />
              </circle>
              <circle
                cx="22"
                cy="22"
                r={1}
              >
                <animate
                  attributeName="r"
                  begin="-0.9s"
                  calcMode="spline"
                  dur="1.8s"
                  keySplines="0.165, 0.84, 0.44, 1"
                  keyTimes="0; 1"
                  repeatCount="indefinite"
                  values="1; 20"
                />
                <animate
                  attributeName="strokeOpacity"
                  begin="-0.9s"
                  calcMode="spline"
                  dur="1.8s"
                  keySplines="0.3, 0.61, 0.355, 1"
                  keyTimes="0; 1"
                  repeatCount="indefinite"
                  values="1; 0"
                />
              </circle>
            </g>
          </svg>
        </div>
      </div>
    </div>
    <span
      className="mega-octicon octicon-home"
      data-testid="home-icon"
      onClick={[Function]}
    />
  </div>
  <div
    className="sc-ksdxgE dJdmIR"
    data-testid="no-results"
  >
    <svg
      fill="currentColor"
      height="2.5em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="2.5em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
      />
    </svg>
    <h2>
      We couldn’t find any matching gists.
    </h2>
  </div>
  <div
    className="sc-hBUSln kkAFFm"
    data-testid="app-footer"
  >
    <p>
      Proudly developed by
      <a
        className="sc-fotOHu bxLnRr"
        href="mailto:vipul.chawda@outlook.com"
        title="You can reach out to me at vipul.chawda@outlook.com"
      >
        <span>
          Vipul Chawda
        </span>
      </a>
    </p>
    © 2021 All rights reserved.
  </div>
</div>
`;
